trl/trainer/stepwise_reward_trainer.py

```diff
@@ -0,0 +1,255 @@
+# Copyright 2023 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import inspect
+import warnings
+from collections import defaultdict
+from dataclasses import FrozenInstanceError, replace
+from functools import wraps
+from typing import Any, Callable, Dict, List, Optional, Tuple, Union
+
+import pandas as pd
+import torch
+import torch.nn as nn
+from accelerate import PartialState
+from accelerate.utils import gather_object
+from datasets import Dataset
+from transformers import (
+    DataCollator,
+    PreTrainedModel,
+    PreTrainedTokenizerBase,
+    Trainer,
+    DataCollatorForTokenClassification,
+)
+from transformers.trainer_callback import TrainerCallback
+from transformers.trainer_pt_utils import nested_detach
+from transformers.trainer_utils import EvalPrediction
+from transformers.utils import is_peft_available
+
+from ..data_utils import maybe_apply_chat_template
+from .stepwise_reward_config import StepwiseRewardConfig
+from .utils import (
+    compute_accuracy,
+    trl_sanitze_kwargs_for_tagging,
+)
+
+
+if is_peft_available():
+    from peft import PeftModel, get_peft_model, prepare_model_for_kbit_training
+
+
+def _tokenize(
+    batch: Dict[str, List[Any]], tokenizer: PreTrainedTokenizerBase, max_length: int, post_step_separator: str
+) -> Dict[str, List[Any]]:
+    """Tokenize a batch from a stepwise preference modeling dataset."""
+    new_examples = {"input_ids": [], "attention_mask": [], "labels": []}
+
+    post_step_tokens = tokenizer.encode(post_step_separator, add_special_tokens=False)
+
+    for prompt, steps, labels in zip(batch["prompt"], batch["stepwise_completion"], batch["stepwise_labels"]):
+        if len(steps) != len(labels):
+            raise NotImplementedError("StepwiseRewardTrainer does not support training with unlabeled steps.")
```

Comment:
Indeed, a `ValueError` makes more sense here, since I am also not aware of any cases where training on unlabeled steps would make sense. In all cases, if the user wants to train with an unlabeled step, he could just set the label to -100.